File: src/actions/builder/builder.rs
use anyhow::{anyhow, bail, Context, Result};
use std::{
    collections::HashSet,
    env,
    fs::{self, File},
    io,
    path::{Path, PathBuf},
    process::Command,
    thread,
};
use walkdir::WalkDir;
use zip::write::SimpleFileOptions;

#[derive(Debug, Clone, Copy)]
enum Platform {
    Modrinth,
    Curseforge,
}

impl Platform {
    fn from_suffix(s: &str) -> Option<Self> {
        match s {
            "mr" => Some(Platform::Modrinth),
            "cf" => Some(Platform::Curseforge),
            _ => None,
        }
    }

    fn short(self) -> &'static str {
        match self {
            Platform::Modrinth => "mr",
            Platform::Curseforge => "cf",
        }
    }

    fn cli(self) -> &'static str {
        match self {
            Platform::Modrinth => "modrinth",
            Platform::Curseforge => "curseforge",
        }
    }

    fn ext(self) -> &'static str {
        match self {
            Platform::Modrinth => "mrpack",
            Platform::Curseforge => "zip",
        }
    }
}

fn main() -> Result<()> {
    let args: Vec<String> = env::args().collect();
    let short_sha = args
        .get(1)
        .ok_or_else(|| anyhow!("usage: builder <short-sha>"))?
        .clone();

    let repo_root = env::current_dir().context("failed to get current directory")?;
    let artifacts_dir = repo_root.join("artifacts");
    fs::create_dir_all(&artifacts_dir)
        .with_context(|| format!("failed to create {}", artifacts_dir.display()))?;

    let changed = detect_changed_targets().context("failed to detect changed targets")?;

    if changed.is_empty() {
        println!("no packs detected in git diff.");
        return Ok(());
    }

    for (category, pack_id) in &changed {
        match category.as_str() {
            "modpacks" => build_modpack(pack_id, &short_sha, &artifacts_dir)
                .with_context(|| format!("modpack '{pack_id}' failed"))?,
            "datapacks" => build_datapack(pack_id, &short_sha, &artifacts_dir)
                .with_context(|| format!("datapack '{pack_id}' failed"))?,
            other => println!("category '{other}' does not require a build."),
        }
    }

    println!("all builds completed successfully.");
    Ok(())
}

fn detect_changed_targets() -> Result<HashSet<(String, String)>> {
    println!("detecting changed files...");
    let output = Command::new("git")
        .args(["diff-tree", "--no-commit-id", "--name-only", "-r", "HEAD"])
        .output()
        .context("failed to invoke git")?;

    if !output.status.success() {
        bail!(
            "git diff-tree failed: {}",
            String::from_utf8_lossy(&output.stderr)
        );
    }

    let stdout = String::from_utf8_lossy(&output.stdout);
    let mut targets = HashSet::new();

    for line in stdout.lines() {
        if line.is_empty() || line.starts_with("external/") || line.starts_with(".actions/") {
            continue;
        }
        let mut parts = line.splitn(3, '/');
        if let (Some(cat), Some(pack)) = (parts.next(), parts.next()) {
            targets.insert((cat.to_string(), pack.to_string()));
        }
    }

    Ok(targets)
}

fn build_modpack(pack_id: &str, sha: &str, artifacts_dir: &Path) -> Result<()> {
    println!("building modpack: {pack_id}");
    let pack_dir = PathBuf::from("modpacks").join(pack_id);

    let manifest_path = pack_dir.join("manifest.json");
    let manifest: serde_json::Value = {
        let file = File::open(&manifest_path)
            .with_context(|| format!("failed to open {}", manifest_path.display()))?;
        serde_json::from_reader(file)
            .with_context(|| format!("invalid JSON in {}", manifest_path.display()))?
    };
    let p_ver = manifest["version"]
        .as_str()
        .ok_or_else(|| anyhow!("missing 'version' in {}", manifest_path.display()))?
        .to_string();

    let mut jobs: Vec<(Platform, PathBuf, String)> = Vec::new();
    let entries = fs::read_dir(&pack_dir)
        .with_context(|| format!("failed to read {}", pack_dir.display()))?;

    for entry in entries {
        let entry = entry.context("failed to read directory entry")?;
        let path = entry.path();
        if !path.is_dir() {
            continue;
        }
        let Some(dir_name) = path.file_name().and_then(|s| s.to_str()) else {
            continue;
        };
        let Some((mc_ver, suffix)) = dir_name.rsplit_once('-') else {
            continue;
        };
        let Some(platform) = Platform::from_suffix(suffix) else {
            continue;
        };
        let mc_ver = mc_ver.to_string();
        jobs.push((platform, path, mc_ver));
    }

    if jobs.is_empty() {
        bail!("no valid version dirs (expected '{{mc_ver}}-mr' or '{{mc_ver}}-cf') for {pack_id}");
    }

    let pack_id_owned = pack_id.to_string();
    let sha_owned = sha.to_string();
    let artifacts_dir_owned = artifacts_dir.to_path_buf();

    let mut handles = Vec::new();
    for (platform, target_path, mc_ver) in jobs {
        let pack_id = pack_id_owned.clone();
        let sha = sha_owned.clone();
        let p_ver = p_ver.clone();
        let artifacts_dir = artifacts_dir_owned.clone();

        handles.push(thread::spawn(move || -> Result<()> {
            let output_name = format!(
                "{}-{}-{}-{}-{}.{}",
                pack_id,
                mc_ver,
                platform.short(),
                p_ver,
                sha,
                platform.ext()
            );
            let output_path = artifacts_dir.join(&output_name);
            let out_str = output_path
                .to_str()
                .ok_or_else(|| anyhow!("non-UTF8 output path"))?;

            let export = Command::new("packwiz")
                .args([platform.cli(), "export", "--output", out_str])
                .current_dir(&target_path)
                .status()
                .context("failed to invoke packwiz export")?;
            if !export.success() {
                bail!("packwiz export failed for {}", target_path.display());
            }

            println!("exported {output_name}");
            Ok(())
        }));
    }

    let mut errors = Vec::new();
    for h in handles {
        match h.join() {
            Ok(Ok(())) => {}
            Ok(Err(e)) => errors.push(e),
            Err(_) => errors.push(anyhow!("export thread panicked")),
        }
    }

    if !errors.is_empty() {
        for e in &errors {
            eprintln!("error: {e:#}");
        }
        bail!("{} export(s) failed for {pack_id}", errors.len());
    }

    Ok(())
}

fn build_datapack(pack_id: &str, sha: &str, artifacts_dir: &Path) -> Result<()> {
    println!("zipping datapack: {pack_id}");
    let src = PathBuf::from("datapacks").join(pack_id);
    let dest = artifacts_dir.join(format!("{pack_id}-{sha}.zip"));
    zip_dir(&src, &dest).with_context(|| format!("failed to zip {}", src.display()))
}

fn zip_dir(src: &Path, dest: &Path) -> Result<()> {
    let file = File::create(dest)
        .with_context(|| format!("failed to create {}", dest.display()))?;
    let mut zip = zip::ZipWriter::new(file);
    let options =
        SimpleFileOptions::default().compression_method(zip::CompressionMethod::Deflated);

    for entry in WalkDir::new(src) {
        let entry = entry.context("walkdir error")?;
        let path = entry.path();
        let name = path.strip_prefix(src).context("strip_prefix failed")?;

        if path.is_file() {
            zip.start_file(name.to_string_lossy(), options)?;
            let mut f = File::open(path)
                .with_context(|| format!("failed to open {}", path.display()))?;
            io::copy(&mut f, &mut zip)
                .with_context(|| format!("failed to write {} to zip", path.display()))?;
        } else if !name.as_os_str().is_empty() {
            zip.add_directory(name.to_string_lossy(), options)?;
        }
    }
    zip.finish()?;
    Ok(())
}

File: src/actions/builder/builder.go
package main

import (
	"archive/zip"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
)

type platform struct {
	short string
	cli   string
	ext   string
}

var (
	modrinth   = platform{short: "mr", cli: "modrinth", ext: "mrpack"}
	curseforge = platform{short: "cf", cli: "curseforge", ext: "zip"}
)

func platformFromSuffix(s string) (platform, bool) {
	switch s {
	case "mr":
		return modrinth, true
	case "cf":
		return curseforge, true
	default:
		return platform{}, false
	}
}

type variant struct {
	MCVersion string `json:"mc_version"`
	ID        string `json:"id,omitempty"`
}

type manifest struct {
	Version  string    `json:"version"`
	Variants []variant `json:"variants,omitempty"`
}

func main() {
	if len(os.Args) < 2 {
		fail("usage: builder <short-sha>")
	}
	shortSHA := os.Args[1]

	repoRoot, err := os.Getwd()
	if err != nil {
		fail(fmt.Sprintf("failed to get current directory: %v", err))
	}
	artifactsDir := filepath.Join(repoRoot, "artifacts")
	if err := os.MkdirAll(artifactsDir, 0o755); err != nil {
		fail(fmt.Sprintf("failed to create %s: %v", artifactsDir, err))
	}

	changed, err := detectChangedTargets()
	if err != nil {
		fail(fmt.Sprintf("failed to detect changed targets: %v", err))
	}
	if len(changed) == 0 {
		fmt.Println("no packs detected in git diff.")
		return
	}

	for _, t := range changed {
		switch t.category {
		case "modpacks":
			if err := buildModpack(t.pack, shortSHA, artifactsDir); err != nil {
				fail(fmt.Sprintf("modpack '%s' failed: %v", t.pack, err))
			}
		case "datapacks", "resourcepacks":
			if err := buildZipPack(t.category, t.pack, shortSHA, artifactsDir); err != nil {
				fail(fmt.Sprintf("%s '%s' failed: %v", t.category, t.pack, err))
			}
		default:
			fmt.Printf("category '%s' does not require a build.\n", t.category)
		}
	}

	fmt.Println("all builds completed successfully.")
}

type target struct {
	category string
	pack     string
}

func detectChangedTargets() ([]target, error) {
	fmt.Println("detecting changed files...")
	out, err := exec.Command("git", "diff-tree", "--no-commit-id", "--name-only", "-r", "HEAD").Output()
	if err != nil {
		return nil, fmt.Errorf("failed to invoke git: %w", err)
	}

	seen := make(map[target]struct{})
	var targets []target

	for _, line := range strings.Split(string(out), "\n") {
		if line == "" || strings.HasPrefix(line, "external/") || strings.HasPrefix(line, ".actions/") {
			continue
		}
		parts := strings.SplitN(line, "/", 3)
		if len(parts) < 2 {
			continue
		}
		t := target{category: parts[0], pack: parts[1]}
		if _, ok := seen[t]; !ok {
			seen[t] = struct{}{}
			targets = append(targets, t)
		}
	}
	return targets, nil
}

func readManifest(path string) (*manifest, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to open %s: %w", path, err)
	}
	var m manifest
	if err := json.Unmarshal(data, &m); err != nil {
		return nil, fmt.Errorf("invalid JSON in %s: %w", path, err)
	}
	return &m, nil
}

func subdirKeys(m *manifest) []string {
	if len(m.Variants) == 0 {
		return nil
	}
	keys := make([]string, 0, len(m.Variants))
	for _, v := range m.Variants {
		if v.ID != "" {
			keys = append(keys, v.ID)
		} else {
			keys = append(keys, v.MCVersion)
		}
	}
	return keys
}

type exportJob struct {
	plat   platform
	dir    string
	subKey string
}

func buildModpack(packID, sha string, artifactsDir string) error {
	fmt.Printf("building modpack: %s\n", packID)
	packDir := filepath.Join("modpacks", packID)

	m, err := readManifest(filepath.Join(packDir, "manifest.json"))
	if err != nil {
		return err
	}
	if m.Version == "" {
		return fmt.Errorf("missing 'version' in %s/manifest.json", packDir)
	}

	var jobs []exportJob
	keys := subdirKeys(m)

	if keys != nil {
		for _, key := range keys {
			for _, plat := range []platform{modrinth, curseforge} {
				dir := filepath.Join(packDir, key+"-"+plat.short)
				if info, err := os.Stat(dir); err == nil && info.IsDir() {
					jobs = append(jobs, exportJob{plat: plat, dir: dir, subKey: key})
				}
			}
		}
	} else {
		entries, err := os.ReadDir(packDir)
		if err != nil {
			return fmt.Errorf("failed to read %s: %w", packDir, err)
		}
		for _, e := range entries {
			if !e.IsDir() {
				continue
			}
			name := e.Name()
			idx := strings.LastIndex(name, "-")
			if idx < 0 {
				continue
			}
			key, suffix := name[:idx], name[idx+1:]
			plat, ok := platformFromSuffix(suffix)
			if !ok {
				continue
			}
			jobs = append(jobs, exportJob{plat: plat, dir: filepath.Join(packDir, name), subKey: key})
		}
	}

	if len(jobs) == 0 {
		return fmt.Errorf("no valid version dirs (expected '{key}-mr' or '{key}-cf') for %s", packID)
	}

	var wg sync.WaitGroup
	errCh := make(chan error, len(jobs))

	for _, j := range jobs {
		wg.Add(1)
		go func(j exportJob) {
			defer wg.Done()
			outputName := fmt.Sprintf("%s-%s-%s-%s-%s.%s",
				packID, j.subKey, j.plat.short, m.Version, sha, j.plat.ext)
			outputPath := filepath.Join(artifactsDir, outputName)

			cmd := exec.Command("packwiz", j.plat.cli, "export", "--output", outputPath)
			cmd.Dir = j.dir
			if out, err := cmd.CombinedOutput(); err != nil {
				errCh <- fmt.Errorf("packwiz export failed for %s: %v\n%s", j.dir, err, out)
				return
			}
			fmt.Printf("exported %s\n", outputName)
		}(j)
	}

	wg.Wait()
	close(errCh)

	var errs []error
	for e := range errCh {
		errs = append(errs, e)
	}
	if len(errs) > 0 {
		for _, e := range errs {
			fmt.Fprintf(os.Stderr, "error: %v\n", e)
		}
		return fmt.Errorf("%d export(s) failed for %s", len(errs), packID)
	}
	return nil
}

func buildZipPack(category, packID, sha, artifactsDir string) error {
	fmt.Printf("zipping %s: %s\n", category, packID)
	packDir := filepath.Join(category, packID)

	versionDir, version, err := findSingleVersionDir(packDir)
	if err != nil {
		return err
	}

	dest := filepath.Join(artifactsDir, fmt.Sprintf("%s-%s-%s.zip", packID, version, sha))
	return zipContents(versionDir, dest)
}

func findSingleVersionDir(packDir string) (dir string, version string, err error) {
	entries, err := os.ReadDir(packDir)
	if err != nil {
		return "", "", fmt.Errorf("failed to read %s: %w", packDir, err)
	}
	var versionDirs []string
	for _, e := range entries {
		if e.IsDir() {
			versionDirs = append(versionDirs, e.Name())
		}
	}
	switch len(versionDirs) {
	case 0:
		return "", "", fmt.Errorf("no version directory found in %s (expected %s/{version}/)", packDir, packDir)
	case 1:
		return filepath.Join(packDir, versionDirs[0]), versionDirs[0], nil
	default:
		return "", "", fmt.Errorf("expected exactly one version directory in %s, found %d: %v", packDir, len(versionDirs), versionDirs)
	}
}

func zipContents(src, dest string) error {
	f, err := os.Create(dest)
	if err != nil {
		return fmt.Errorf("failed to create %s: %w", dest, err)
	}
	defer f.Close()

	zw := zip.NewWriter(f)
	defer zw.Close()

	return filepath.Walk(src, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}
		rel, err := filepath.Rel(src, path)
		if err != nil {
			return err
		}
		if rel == "." {
			return nil
		}
		zipName := filepath.ToSlash(rel)
		if info.IsDir() {
			_, err := zw.Create(zipName + "/")
			return err
		}
		w, err := zw.Create(zipName)
		if err != nil {
			return err
		}
		in, err := os.Open(path)
		if err != nil {
			return fmt.Errorf("failed to open %s: %w", path, err)
		}
		defer in.Close()
		_, err = io.Copy(w, in)
		return err
	})
}

func fail(msg string) {
	fmt.Fprintf(os.Stderr, "::error::%s\n", msg)
	os.Exit(1)
}
